<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Aggregate data in SFE using geometry — aggregate,SpatialFeatureExperiment-method • SpatialFeatureExperiment</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Aggregate data in SFE using geometry — aggregate,SpatialFeatureExperiment-method"><meta name="description" content="Gene expression and numeric columns of colData will be aggregated with
the function specified in FUN, according to another geometry supplied
and a geometry predicate (such as st_intersects). For example, when
the predicate is st_intersects and a spatial grid is used to
aggregate, then the data associated with all cells that intersect with each
grid cell will be aggregated with FUN, such as mean or
sum. The categorical columns will be collected into list columns, and
logical columns will be converted into numeric before applying FUN."><meta property="og:description" content="Gene expression and numeric columns of colData will be aggregated with
the function specified in FUN, according to another geometry supplied
and a geometry predicate (such as st_intersects). For example, when
the predicate is st_intersects and a spatial grid is used to
aggregate, then the data associated with all cells that intersect with each
grid cell will be aggregated with FUN, such as mean or
sum. The categorical columns will be collected into list columns, and
logical columns will be converted into numeric before applying FUN."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialFeatureExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/seurat_sfe_coerce.html">Seurat_to_SFE</a></li>
    <li><a class="dropdown-item" href="../articles/SFE.html">Introduction to the SpatialFeatureExperiment class</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/SpatialFeatureExperiment/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Aggregate data in SFE using geometry</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/SpatialFeatureExperiment/blob/documentation/R/aggregate.R" class="external-link"><code>R/aggregate.R</code></a></small>
      <div class="d-none name"><code>aggregate-SpatialFeatureExperiment-method.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Gene expression and numeric columns of <code>colData</code> will be aggregated with
the function specified in <code>FUN</code>, according to another geometry supplied
and a geometry predicate (such as <code>st_intersects</code>). For example, when
the predicate is <code>st_intersects</code> and a spatial grid is used to
aggregate, then the data associated with all cells that intersect with each
grid cell will be aggregated with <code>FUN</code>, such as <code>mean</code> or
<code>sum</code>. The categorical columns will be collected into list columns, and
logical columns will be converted into numeric before applying <code>FUN</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'SpatialFeatureExperiment'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  FUN <span class="op">=</span> <span class="va">sum</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  rowGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellsize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  square <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  flat_topped <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  new_geometry_name <span class="op">=</span> <span class="st">"bins"</span>,</span>
<span>  join <span class="op">=</span> <span class="va">st_intersects</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An SFE object to be aggregated.</p></dd>


<dt id="arg-by">by<a class="anchor" aria-label="anchor" href="#arg-by"></a></dt>
<dd><p>A <code>sf</code> data frame whose geometry column is used for
aggregation or <code>sfc</code> or for multiple samples a list of <code>sfc</code>
whose names are the sample IDs. For multiple samples, the <code>sf</code> data
frame must have a column <code>sample_id</code> to indicate which geometry for
which sample. This argument is optional if <code>cellsize</code> is specified.</p></dd>


<dt id="arg-fun">FUN<a class="anchor" aria-label="anchor" href="#arg-fun"></a></dt>
<dd><p>Function to aggregate the numerical columns in <code>colData</code> and
the gene count matrix. This can be <code>sum</code>, <code>mean</code>, or any function
that takes a numeric matrix as input and returns a numeric vector whose
length is same as the number of rows in the input matrix, such as
<code>rowMedians</code>. See package <code>matrixStats</code>. Depending on the
function used for aggregation, numeric columns of <code>colData</code> may need
to be interpreted differently after aggregation. Aggregation is not done
when aggregating by transcript spots in <code>rowGeometry</code>. When it's sum
or mean, matrix multiplication is used for aggregation rather than calling
the sum or mean function itself; this is much faster than looping through
the bins and calling the function on each of them.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>Which samples to aggregate, defaults to "all".</p></dd>


<dt id="arg-colgeometryname">colGeometryName<a class="anchor" aria-label="anchor" href="#arg-colgeometryname"></a></dt>
<dd><p>Which <code>colGeometry</code> to spatially aggregate the
data, by default the first one.</p></dd>


<dt id="arg-rowgeometryname">rowGeometryName<a class="anchor" aria-label="anchor" href="#arg-rowgeometryname"></a></dt>
<dd><p>Which <code>rowGeometry</code> to spatially aggregate</p></dd>


<dt id="arg-cellsize">cellsize<a class="anchor" aria-label="anchor" href="#arg-cellsize"></a></dt>
<dd><p>numeric of length 1 or 2 with target cellsize: for square or rectangular cells the width and height, for hexagonal cells the distance between opposite edges (edge length is cellsize/sqrt(3)). A length units object can be passed, or an area unit object with area size of the square or hexagonal cell.</p></dd>


<dt id="arg-square">square<a class="anchor" aria-label="anchor" href="#arg-square"></a></dt>
<dd><p>logical; if <code>FALSE</code>, create hexagonal grid</p></dd>


<dt id="arg-flat-topped">flat_topped<a class="anchor" aria-label="anchor" href="#arg-flat-topped"></a></dt>
<dd><p>logical; if <code>TRUE</code> generate flat topped hexagons, else generate pointy topped</p></dd>


<dt id="arg-new-geometry-name">new_geometry_name<a class="anchor" aria-label="anchor" href="#arg-new-geometry-name"></a></dt>
<dd><p>Name to give to the new <code>colGeometry</code> in the
output. Defaults to "bins".</p></dd>


<dt id="arg-join">join<a class="anchor" aria-label="anchor" href="#arg-join"></a></dt>
<dd><p>logical spatial predicate function to use if <code>by</code> is a simple features object or geometry; see <a href="https://r-spatial.github.io/sf/reference/st_join.html" class="external-link">st_join</a></p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code>BiocParallelParam</code> object specifying parallel
computing when aggregating data with functions other than sum and mean.
Defaults to <code>SerialParam()</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An SFE object with <code>colGeometry</code> the same as the geometry
  specified in <code>by</code> or same as the grid specified in <code>cellsize</code>.
  <code>rowGeometries</code> and <code>rowData</code> remain the same as in the input
  <code>x</code>. <code>reducedDims</code>, <code>localResults</code>, <code>colFeatureData</code>
  (and its <code>colGeometry</code>, <code>annotGeometry</code>, and <code>reducedDim</code>
  counterparts), and <code>spatialGraphs</code> are dropped because those results
  no longer apply after aggregation.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For smFISH-based data where the transcript spots are available, the
transcript spots can be used instead of cells to aggregate the gene count
matrix, in which case all assays other than <code>counts</code> will be dropped and
<code>FUN</code> only applies to <code>colData</code> because the transcript spots are
simply counted.</p>
<p>What this function does is similar to
<a href="https://github.com/JEFworks-Lab/SEraster" class="external-link">SEraster</a> but more general
because any geometry and more aggregation function can be used, not just
regular grids, and the aggregation can be performed on the transcript spots.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>For developers: When debugging this function after calling
  <code>devtools::load_all(".")</code>, you may get an error that comes from S3
  dispatch of <code>aggregate.Vector</code> from the <code>S4Vectors</code> package. When
  that happens, either restart the R session, or run
  <code>setGeneric("aggregate", function(x, ...)
  standardGeneric("aggregate"))</code> in the console to make an S4 generic as done
  in the <code>terra</code> package to prioritize S4 dispatch.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># example code</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

