<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read 10X Xenium output as SpatialFeatureExperiment — readXenium • SpatialFeatureExperiment</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read 10X Xenium output as SpatialFeatureExperiment — readXenium"><meta name="description" content="This function reads the standard 10X Xenium output into an SFE object."><meta property="og:description" content="This function reads the standard 10X Xenium output into an SFE object."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialFeatureExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/seurat_sfe_coerce.html">Seurat_to_SFE</a></li>
    <li><a class="dropdown-item" href="../articles/SFE.html">Introduction to the SpatialFeatureExperiment class</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/SpatialFeatureExperiment/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read 10X Xenium output as SpatialFeatureExperiment</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/SpatialFeatureExperiment/blob/documentation/R/read.R" class="external-link"><code>R/read.R</code></a></small>
      <div class="d-none name"><code>readXenium.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reads the standard 10X Xenium output into an SFE object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readXenium</span><span class="op">(</span></span>
<span>  <span class="va">data_dir</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"sample01"</span>,</span>
<span>  min_area <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  image <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"morphology_focus"</span>, <span class="st">"morphology_mip"</span><span class="op">)</span>,</span>
<span>  segmentations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell"</span>, <span class="st">"nucleus"</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"symbol"</span><span class="op">)</span>,</span>
<span>  flip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geometry"</span>, <span class="st">"image"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  max_flip <span class="op">=</span> <span class="st">"50 MB"</span>,</span>
<span>  filter_counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add_molecules <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min_phred <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  file_out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"tx_spots.parquet"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-dir">data_dir<a class="anchor" aria-label="anchor" href="#arg-data-dir"></a></dt>
<dd><p>Top level output directory.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>A <code>character</code> sample identifier, which matches the
<code>sample_id</code> in <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></code>. The <code>sample_id</code> will also
be stored in a new column in <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-colData.html" class="external-link">colData</a></code>, if not already present.
Default = <code>sample01</code>.</p></dd>


<dt id="arg-min-area">min_area<a class="anchor" aria-label="anchor" href="#arg-min-area"></a></dt>
<dd><p>Minimum cell area in square microns or pixel units (eg for CosMX).
Anything smaller will be considered artifact or debris and removed.
Default to `NULL`, ie no filtering of polygons.</p></dd>


<dt id="arg-image">image<a class="anchor" aria-label="anchor" href="#arg-image"></a></dt>
<dd><p>Which image(s) to load, can be "morphology_mip",
"morphology_focus" or both. Note that in Xenium Onboarding Analysis (XOA)
v2, there is no longer "morphology_mip" and "morphology_focus" is a
directory with 4 images corresponding to 4 channels: DAPI, "Cadherin", 18S,
and Vimentin. So this argument is ignored for XOA v2.</p></dd>


<dt id="arg-segmentations">segmentations<a class="anchor" aria-label="anchor" href="#arg-segmentations"></a></dt>
<dd><p>Which segmentation outputs to read, can be "cell",
"nucleus", or both.</p></dd>


<dt id="arg-row-names">row.names<a class="anchor" aria-label="anchor" href="#arg-row-names"></a></dt>
<dd><p>String specifying whether to use Ensembl IDs ("id") or gene
symbols ("symbol") as row names. If using symbols, the Ensembl ID will be
appended to disambiguate in case the same symbol corresponds to multiple
Ensembl IDs. Always "symbol" if `add_molecules = TRUE` because only gene
symbols are used in the transcript spot files.</p></dd>


<dt id="arg-flip">flip<a class="anchor" aria-label="anchor" href="#arg-flip"></a></dt>
<dd><p>To flip the image, geometry coordinates, or none. Because the
image has the origin at the top left while the geometry has origin at the
bottom left, one of them needs to be flipped for them to match. If one of
them is already flipped, then use "none". The image will not be flipped if
it's GeoTIFF.</p></dd>


<dt id="arg-max-flip">max_flip<a class="anchor" aria-label="anchor" href="#arg-max-flip"></a></dt>
<dd><p>Maximum size of the image allowed to flip the image. Because
the image will be loaded into memory to be flipped. If the image is larger
than this size then the coordinates will be flipped instead.</p></dd>


<dt id="arg-filter-counts">filter_counts<a class="anchor" aria-label="anchor" href="#arg-filter-counts"></a></dt>
<dd><p>Logical, whether to keep cells with counts <code>&gt; 0</code>.</p></dd>


<dt id="arg-add-molecules">add_molecules<a class="anchor" aria-label="anchor" href="#arg-add-molecules"></a></dt>
<dd><p>Logical, whether to add transcripts coordinates to an
object.</p></dd>


<dt id="arg-min-phred">min_phred<a class="anchor" aria-label="anchor" href="#arg-min-phred"></a></dt>
<dd><p>Minimum Phred score to keep spot. By default 20, the
conventional threshold indicating "acceptable", meaning that there's 1<!-- % -->
chance that the spot was decoded in error.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code>BiocParallelParam</code> object specifying parallel
  processing backend and number of threads to use for parallelizable tasks:</p><ol><li><p>To load cell segmentation from HDF5 files from different
  fields of view (FOVs) with multiple cores. A progress bar can be configured
  in the <code>BiocParallelParam</code> object. When there are numerous
  FOVs, reading in the geometries can be time consuming, so we recommend
  using a server and larger number of threads. This argument is not used if
  <code>use_cellpose = TRUE</code> and the parquet file is present.</p></li>
<li><p>To get the largest piece and see if it's larger than <code>min_area</code>
  when there are multiple pieces in the cell segmentation for one cell.</p></li>
</ol></dd>


<dt id="arg-file-out">file_out<a class="anchor" aria-label="anchor" href="#arg-file-out"></a></dt>
<dd><p>Name of file to save the geometry or raster to disk.
Especially when the geometries are so large that it's unwieldy to load
everything into memory. If this file (or directory for multiple files)
already exists, then the existing file(s) will be read, skipping the
processing. When writing the file, extensions supplied are ignored and
extensions are determined based on `dest`.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An SFE object. If reading segmentations, the cell or nuclei
  segmentation will be saved to `cell_boundaries_sf.parquet` and
  `nucleus_boundaries_sf.parquet` respectively in `data.dir` so next time the
  boundaries can be read much more quickly. If reading transcript spots
  (`add_molecules = TRUE`), then the reformatted transcript spots are saved
  to file specified in the `file_out` argument, which is by default
  `tx_spots.parquet` in the same directory as the rest of the data. If images
  are present, then the images will be of the <code>BioFormatsImage</code> class
  and not loaded into memory until necessary in later operations.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>Sometimes when reading images, you will see this error the first time:
'java.lang.NullPointerException: Cannot invoke
"loci.formats.DimensionSwapper.setMetadataFiltered(boolean)" because
"RBioFormats.reader" is null'. See this issue https://github.com/aoles/RBioFormats/issues/42
Rerun the code and it should work the second time.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/RBioFormats" class="external-link">RBioFormats</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="va">fp</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The downloaded files are in /tmp/RtmpgNzKQ4/file830670f95499/xenium2 </span>
<span class="r-in"><span><span class="co"># RBioFormats issue</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">readXenium</span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell and nucleus segmentations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Making POLYGON cell geometries</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sanity checks on cell segmentation polygons:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; ..found 132 cells with (nested) polygon lists</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; ..applying filtering</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Checking polygon validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Saving geometries to parquet files</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading h5 gene count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading transcript coordinates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Converting transcript spots to geometry</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Writing reformatted transcript spots to disk</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">readXenium</span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Preprocessed sf segmentations found</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell and nucleus segmentations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading h5 gene count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading transcript coordinates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">dir_use</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

